### ISSUE TITLE
Merge Request diff comment support for GitLab

### ISSUE THREAD

**Issue by connorshea** [danger/danger#522]
This is a tracking issue for adding support for inline comments to Danger's GitLab integration.

Relevant GitLab Issue: https://gitlab.com/gitlab-org/gitlab-ce/issues/16550

Relevant GitLab MR: https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/4474

Support for this will also need to be added to the gitlab gem.

**Comment by jk** [comment]
Just came across this issue. Looks like the linked gitlab issues are closed/merged and the API [describes](https://docs.gitlab.com/ee/api/commits.html#post-comment-to-commit) adding inline comments to commits (don't see adding comments/notes on MR here…). So this feature might become possible now?

**Comment by born4new** [comment]
I would love that feature...If you are interested, I can go ahead an make a PR for it?

Waiting on your feedback before starting anything else...

**Comment by ZJvandeWeg** [comment]
@born4new Danger depends on https://github.com/NARKOZ/gitlab/ and I didn't spot any discussion related endpoints in https://github.com/NARKOZ/gitlab/tree/master/lib/gitlab/client, I think thats a required first step?

**Comment by born4new** [comment]
It sure it @ZJvandeWeg :). I'll propose to implement the endpoint on there as well.

**Comment by saurabhnanda** [comment]
I'm integrating Danger into our Gitlab project and stumbled upon this issue. This is waiting for someone to implement it, now that the `gitlab` gem has support for MR discussions, right?

**Comment by NARKOZ** [comment]
I've just released the gitlab gem v4.4.0 with MR discussions feature ✌️

**Comment by saurabhnanda** [comment]
Thanks @NARKOZ 👍 

IIUC, `danger` gem now needs to be upgraded to take advantage of this new functionality, right?

**Comment by orta** [comment]
depending on what version number the new one is, it might just [be fine as is](https://github.com/danger/danger-gitlab-gem/blob/master/danger-gitlab.gemspec#L17) - though I'd prefer to make a new lower bounds for new functionality 👍

**Comment by dimazen** [comment]
Any updates on it?

**Comment by orta** [comment]
Nope, this issue is where someone would probably post updates - you're welcome to look at it

**Comment by dimazen** [comment]
@orta sorry for disturbing you, but just to confirm:
In order to have inline comments for the GitLab we need to implement in [gitlab.rb:112](https://github.com/danger/danger/blob/master/lib/danger/request_sources/gitlab.rb#L112) logic similar to what Danger current has in [github.rb:143](https://github.com/danger/danger/blob/master/lib/danger/request_sources/github/github.rb#L143)? 
i.e. GitHub uses warnings from Danger to build inline messages, while GitLab simply generates a single comment. 
Am I right?

**Comment by orta** [comment]
Looking at their API, I don't think so - https://docs.gitlab.com/ce/api/commits.html#post-comment-to-commit

In GitHub you're attaching a comment _into_ the diff, so we have to find it's place in there, in gitlab it looks like it just accepts file and line. 

The thing we needed for GitHub was to make sure that the comment would be seen because if the comment didn't live in the range of the diff it wouldn't show ( and would need to go in the main message instead )
